prober/icmp.go

```diff
@@ -243,13 +279,51 @@ func ProbeICMP(ctx context.Context, target string, module config.Module, registr
 
 	rb := make([]byte, 65536)
 	deadline, _ := ctx.Deadline()
-	if err := socket.SetReadDeadline(deadline); err != nil {
+	if icmpConn != nil {
+		err = icmpConn.SetReadDeadline(deadline)
+	} else {
+		err = v4RawConn.SetReadDeadline(deadline)
+	}
+	if err != nil {
 		level.Error(logger).Log("msg", "Error setting socket deadline", "err", err)
 		return
 	}
 	level.Info(logger).Log("msg", "Waiting for reply packets")
 	for {
-		n, peer, err := socket.ReadFrom(rb)
+		var n int
+		var peer net.Addr
+		var err error
+		var hopLimit float64 = -1
```

Comment:
This is still messy, as there's now two hasHopLimit indicators doing different things. It'd be better to name the bool what is actually, namely whether the CM was set.

Reply:
The real issue is that go does not tell me if hopLimit is missing or zero. I opened an issue golang/go#41820.

I renamed the bool to ipv6HopLimitFlagSet and now I use it only for IPv6. If golang fixes the issue, it can be removed in the future to adapt to whatever go uses to inform hopLimit is missing.

Another solution is to assume that hopLimit will never be 0 in a echo-reply. I assumed that hop limit could be zero when it is using the last permitted hop. However, I was wrong as a packet with hopLimit==1 will not get forwarded. The IPv6 echo reply will only be 0 if the target OS is explicitly setting the answer to be 0 (maybe a way to avoid its traffic to be routed?). As it is a very very special case, we might ignore it.

What do you think?

Reply:
Probably best not to make assumptions about the TTL in a tool like this, as if that weird situation does happen you wouldn't want to lead users down the wrong path. It's probably best to use the flag for both protocols.

Reply:
OK. Let's play safe. I reintroduced the bool check for both.